## In Class Assignment April 9th

In [1]:
%%capture
pip install scikit-learn

In [2]:
%%capture
pip install sweetviz

In [3]:
%%capture
pip install xgboost

In [4]:
import pandas as pd
import numpy as np
import sweetviz as sv
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, accuracy_score

c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [5]:
# read in data
credit_card = pd.read_csv("C:\\Users\\donya\\Downloads\\creditcard.csv")
credit_card.head()

,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,PAY_5,...,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default payment next month
0,400000,1,1,2,32,0,0,0,0,0,...,48272,49478,51242,3028,3023,3000,3000,3000,38662,0
1,120000,2,2,2,30,-1,-1,-1,-1,-1,...,1964,1883,1538,3230,3011,1964,1883,1538,1911,0
2,270000,2,2,2,32,0,0,0,0,0,...,94856,86461,83650,1808,69563,2891,2689,3012,2771,0
3,280000,2,2,1,27,0,0,0,0,0,...,257689,193231,191143,11052,9563,15017,5374,5420,6021,0
4,30000,2,1,2,27,0,0,-1,0,0,...,1814,0,0,1000,664,1500,0,0,0,0


#1

In [6]:
# EDA with sweetviz
report = sv.analyze(credit_card)
report.show_html("sweetviz_report.html")

Done! Use 'show' commands to display/save.   |██████████| [100%]   00:01 -> (00:00 left)


Report sweetviz_report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


Highlights from EDA
- Limit balance skewed to the right with a median of $140,000
- Age is skewed to the right with a median of 34 years old. All individuals in this data set are > 18 years old
- All bill amounts are skewed to the right
- The outcome variable, default payment next month, is imbalanced (individuals primarily will not default on their payment next month)

In [7]:
# define X and y
X = credit_card.drop("default payment next month", axis=1)
y = credit_card["default payment next month"]

# split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=321)

#training random forest model (baseline)
rf = RandomForestClassifier(random_state=321)
rf.fit(X_train, y_train)

#evaluating random forest model
y_pred_rf = rf.predict(X_test)
print("Random Forest Classification Report:")
print(classification_report(y_test, y_pred_rf))
print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))

#with cross validation
cv_scores_rf = cross_val_score(rf, X, y, cv=5)
print("Random Forest Cross-Validation Scores:", cv_scores_rf)
print("Random Forest Average CV Score:", np.mean(cv_scores_rf))
print("Random Forest CV Score Standard Deviation:", np.std(cv_scores_rf))

#training xgboost model (baseline)
xgb = XGBClassifier(random_state=321, use_label_encoder=False, eval_metric='logloss')
xgb.fit(X_train, y_train)

#evaluating xgboost model
y_pred_xgb = xgb.predict(X_test)
print("XGBoost Classification Report:")
print(classification_report(y_test, y_pred_xgb))
print("XGBoost Accuracy:", accuracy_score(y_test, y_pred_xgb))

#with cross validation
cv_scores_xgb = cross_val_score(xgb, X, y, cv=5)
print("XGBoost Cross-Validation Scores:", cv_scores_xgb)
print("XGBoost Average CV Score:", np.mean(cv_scores_xgb))
print("XGBoost CV Score Standard Deviation:", np.std(cv_scores_xgb))

Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.95      0.89      3704
           1       0.68      0.40      0.51      1096

    accuracy                           0.82      4800
   macro avg       0.76      0.67      0.70      4800
weighted avg       0.81      0.82      0.80      4800

Random Forest Accuracy: 0.8210416666666667
Random Forest Cross-Validation Scores: [0.81229167 0.81041667 0.81395833 0.81145833 0.81958333]
Random Forest Average CV Score: 0.8135416666666666
Random Forest CV Score Standard Deviation: 0.003235544776386203


c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:15:50] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBoost Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.94      0.89      3704
           1       0.67      0.40      0.50      1096

    accuracy                           0.82      4800
   macro avg       0.75      0.67      0.70      4800
weighted avg       0.80      0.82      0.80      4800

XGBoost Accuracy: 0.8177083333333334


c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:15:50] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:15:51] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:15:51] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:15:51] WARNING: C:\actio

XGBoost Cross-Validation Scores: [0.80979167 0.81083333 0.814375   0.81020833 0.81895833]
XGBoost Average CV Score: 0.8128333333333334
XGBoost CV Score Standard Deviation: 0.00346510381438197


#2
- Both models perform well in terms of accuracy (~ 0.80 for each). However it seems like due to the imbalance in the dataset, the models' high accuracy scores are likely being dominated by correct classifications of individuals who will not default on their payment the next month (outcome = 0). We can see this in the lower recall and precision scores for both models for the outcome = 0 group compared to the outcome = 1 (defaults on payment the next month) group

- The average accuracy scores and standard deviation for the cross-validated models are very similar between both the Random Forest and XGBoost models. We see accuracy scores that are consistently around 0.80 for both models with not much deviation. Due to the imbalanced data, we are likely obtaining folds of data that are primarily representing individuals who do not default on the next month's payment. This may explain the small deviation in accuracy score across the five folds. The model is consistently performing well at classifying our default = 1 cases.

#3
- Yes, as I mentioned in #2, the accuracy score seems to be high due to the model primarily correctly classifying individuals who will not default on next month's payment compared to those who will default. This is likely because of the imbalanced nature of the dataset. There are more individuals who do not actually default in the next month's payment, so these classifications are captured more by our models

#4

In [8]:
#adjust for class imbalance in random forest model
rf_balanced = RandomForestClassifier(random_state=321, class_weight='balanced')
rf_balanced.fit(X_train, y_train)
y_pred_rf_balanced = rf_balanced.predict(X_test)
print("Random Forest (Balanced) Classification Report:")
print(classification_report(y_test, y_pred_rf_balanced))

#adjust for class imbalance in xgboost model
xgb_balanced = XGBClassifier(random_state=321, use_label_encoder=False, eval_metric='logloss', scale_pos_weight=10)
xgb_balanced.fit(X_train, y_train)
y_pred_xgb_balanced = xgb_balanced.predict(X_test)
print("XGBoost (Balanced) Classification Report:")
print(classification_report(y_test, y_pred_xgb_balanced))

Random Forest (Balanced) Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.95      0.89      3704
           1       0.71      0.37      0.49      1096

    accuracy                           0.82      4800
   macro avg       0.77      0.66      0.69      4800
weighted avg       0.81      0.82      0.80      4800



c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:15:57] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBoost (Balanced) Classification Report:
              precision    recall  f1-score   support

           0       0.89      0.62      0.73      3704
           1       0.36      0.74      0.49      1096

    accuracy                           0.65      4800
   macro avg       0.63      0.68      0.61      4800
weighted avg       0.77      0.65      0.68      4800



- The accuracy did not change for the random forest balanced model, but the precision score increased for classification of individuals who will default on the next month's payment. However, the recall score decreased meaning that the model struggled with correctly classifying individuals who default correctly.

- Interestingly, we see accuracy decrease when accounting for class imbalance with XGBoost. The classification report of this model is quite different from the random forest balanced model. In this model, the recall increased substantially but decreased quite a bit as well in terms of precision. This indicates that of individuals who actually defaulted, the model was able to capture 74% (high accuracy) of them but there are many false positives-individuals who were classified as defaulting but they actually did not default (low precision).

#5

In [9]:
#XGBoost model with subsample feature
xgb_subsample = XGBClassifier(random_state=321, use_label_encoder=False, eval_metric='logloss', subsample=0.8)
xgb_subsample.fit(X_train, y_train)
y_pred_xgb_subsample = xgb_subsample.predict(X_test)
print("XGBoost (Subsample) Classification Report:")
print(classification_report(y_test, y_pred_xgb_subsample))

c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:15:57] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBoost (Subsample) Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.94      0.89      3704
           1       0.65      0.40      0.49      1096

    accuracy                           0.81      4800
   macro avg       0.74      0.67      0.69      4800
weighted avg       0.80      0.81      0.80      4800



- The model's accuracy is close to the original XGBoost model without any subsampling

- Both precision and recall have increased for the defaulted = 1 outcome compared to the original XGBoost model. This indicates that the model with subsampling is better able to classify individuals who default on their payments in the next month

#6

In [10]:
#tuning 2 hyperparameters with 2 values for Random Forest using imbalance adjustments
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10]
}
grid_rf = GridSearchCV(estimator=rf_balanced, param_grid=param_grid_rf, cv=5, n_jobs=-1)
grid_rf.fit(X_train, y_train)
print("Best Hyperparameters for Random Forest:", grid_rf.best_params_)
best_rf = grid_rf.best_estimator_
y_pred_best_rf = best_rf.predict(X_test)
print("Tuned Random Forest Classification Report:")
print(classification_report(y_test, y_pred_best_rf))

#tuning 2 hyperparameters with 2 values for XGBoost using imbalance adjustments
param_grid_xgb = {
    'n_estimators': [100, 200],
    'max_depth': [3, 6]
}
grid_xgb = GridSearchCV(estimator=xgb_balanced, param_grid=param_grid_xgb, cv=5, n_jobs=-1)
grid_xgb.fit(X_train, y_train)
print("Best Hyperparameters for XGBoost:", grid_xgb.best_params_)
best_xgb = grid_xgb.best_estimator_
y_pred_best_xgb = best_xgb.predict(X_test)
print("Tuned XGBoost Classification Report:")
print(classification_report(y_test, y_pred_best_xgb))


Best Hyperparameters for Random Forest: {'max_depth': None, 'n_estimators': 100}
Tuned Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.95      0.89      3704
           1       0.71      0.37      0.49      1096

    accuracy                           0.82      4800
   macro avg       0.77      0.66      0.69      4800
weighted avg       0.81      0.82      0.80      4800



c:\Users\donya\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\xgboost\training.py:200: UserWarning: [20:16:34] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


Best Hyperparameters for XGBoost: {'max_depth': 6, 'n_estimators': 200}
Tuned XGBoost Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.73      0.79      3704
           1       0.41      0.64      0.50      1096

    accuracy                           0.71      4800
   macro avg       0.64      0.68      0.65      4800
weighted avg       0.77      0.71      0.73      4800



- The random forest model's accuracy score is the same as the default random forest's accuracy score. However, we see that this model's precision score for default = 1 has increased compared to the default model but also decreased in terms of its recall. Therefore, of everyone the model predicted as defaulting on next month's payment, the model is able to capture a 71% of individuals who actually did default, but still incorrectly classifying those who defaulted as individuals who did not default.

- The accuracy score for the balanced XGBoost model has decreased compared to the default model. The precision score for default = 1 has decreased, while the recall score for this group has increased. This indicates that 64% of individuals are classified correctly as defaulting, but of all the individuals the model predicted would default, only 41% actually defaulted.

- Given the time constraints (and therefore limited time to tune models) of this in-class activity and the fact that each model's performance did not seem to substantially improve from the default models, it is likely that I did not chose the right parameters to tune. There is much more exploration into the vast amount of parameters that I can take to find the optimal model for each model type.

#7

Using the default settings, the random forest classification model performs slightly better than the XGBoost model. If we look at the F1-score of the default = 1 group (takes into account the scores of both precision and recall), we can see that the random forest classification model has the higher score of 0.51 compared to the XGBoost model with an F1-score of 0.50. This is likely because the random forest classification model performs slightly better in terms of precision for this group compared to the XGBoost model (0.68 vs. 0.67). This indicates that the 68% of individuals who are classified as defaulting on next month's payment are actually classified correctly, compared to 67% for the XGBoost model.